libs/workerpool/worker_pool.go

```diff
@@ -0,0 +1,200 @@
+package workerpool
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+
+	sync "github.com/sasha-s/go-deadlock"
+
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+var (
+	ErrWorkerPoolStopped = errors.New("worker-pool has stopped")
+
+	ErrCannotReadResultChannel = errors.New("cannot read result from a channel")
+)
+
+type (
+	// JobSender is an interface that wraps basic job sender methods
+	JobSender interface {
+		Send(ctx context.Context, jobs ...Job) error
+	}
+	// JobReceiver is an interface that wraps basic job receiver methods
+	JobReceiver interface {
+		Receive(ctx context.Context) (Result, error)
+	}
+	// Job is an interface that wraps the job methods that a job must implement to be executed on a worker
+	Job interface {
+		Execute(ctx context.Context) Result
+	}
+	// Result this structure is the result of the job and will be sent to main goroutine via result channel
+	Result struct {
+		Value any
+		Err   error
+	}
+	// OptionFunc is function type for a worker-pool optional functions
+	OptionFunc func(*WorkerPool)
+)
+
+// WithLogger sets a logger to worker-pool using option function
+func WithLogger(logger log.Logger) OptionFunc {
+	return func(p *WorkerPool) {
+		p.logger = logger
+	}
+}
+
+// WithResultCh sets a result channel to worker-pool using option function
+func WithResultCh(resultCh chan Result) OptionFunc {
+	return func(p *WorkerPool) {
+		p.resultCh = resultCh
+	}
+}
+
+// WithJobCh sets a job channel to worker-pool using option function
+func WithJobCh(jobCh chan Job) OptionFunc {
+	return func(p *WorkerPool) {
+		p.jobCh = jobCh
+	}
+}
+
+// WorkerPool is an implementation of a component that allows creating a set of workers
+// to process arbitrary jobs in background
+type WorkerPool struct {
+	initPoolSize int
+	maxPoolSize  int
+	jobCh        chan Job
+	wg           sync.WaitGroup
+	wgMtx        sync.Mutex
+	stopped      atomic.Bool
```

Comment:
no, it is added by purpose, using a channel is inconvenient, because we must care about concurrency since it is possible to create a new done channel in `Reset` method